python/ray/train/_internal/storage.py

```diff
@@ -0,0 +1,154 @@
+import dataclasses
+import os
+from typing import Optional
+
+import pyarrow.fs
+
+from ray.air._internal.uri_utils import is_uri
+from ray.air._internal.remote_storage import get_fs_and_path
+from ray.tune.syncer import Syncer, SyncConfig, _DefaultSyncer
+from ray.tune.result import _get_defaults_results_dir
+
+
+def _use_storage_context() -> bool:
+    # Whether to enable the new simple persistence mode.
+    return bool(int(os.environ.get("RAY_AIR_NEW_PERSISTENCE_MODE", "0")))
+
+
+class StorageContext:
+    """Shared context that holds all paths and storage utilities, passed along from
+    the driver to workers.
+
+    Example:
+        storage_path = "s3://bucket/results"
+        storage_filesystem = S3FileSystem (auto-resolved)
+        storage_cache_path = "~/ray_results"
+        storage_fs_path = "bucket/results"
+
+        experiment_dir_name = "exp_name"
+        # experiment_path = "s3://bucket/results/exp_name"
+        experiment_fs_path = "bucket/results/exp_name"  # Use this path with pyarrow
+        experiment_cache_path = "~/ray_results/exp_name"
+
+        # Only set on workers.
+        trial_dir_name = "trial_dir"
+        # trial_path = "s3://bucket/results/exp_name/trial_dir"
+        trial_fs_path = "bucket/results/exp_name/trial_dir"
+        # trial_cache_path = "~/ray_results/exp_name/trial_dir"
+
+    Internal Usage Examples:
+        construct_checkpoint_fs_path("checkpoint_000001")
+        -> "bucket/results/exp_name/trial_dir/checkpoint_000001"
+
+        pyarrow.fs.copy_files(
+            local_dir,
+            os.path.join(storage.trial_fs_path, "subdir"),
+            destination_filesystem=storage.filesystem
+        )
+    """
+
+    def __init__(
+        self,
+        storage_path: str,
+        sync_config: SyncConfig,
+        experiment_dir_name: str,
+        storage_filesystem: Optional[pyarrow.fs.FileSystem] = None,
+        trial_dir_name: Optional[str] = None,
+    ):
+        self.storage_path: str = storage_path
+        self.storage_cache_path: str = _get_defaults_results_dir()
+        self.experiment_dir_name: str = experiment_dir_name
+        self.trial_dir_name: Optional[str] = trial_dir_name
+        self.sync_config: SyncConfig = dataclasses.replace(sync_config)
+
+        if storage_filesystem:
+            # Custom pyarrow filesystem
+            self.storage_filesystem = storage_filesystem
+            if is_uri(self.storage_path):
+                raise ValueError("TODO")
+            self.storage_fs_path = self.storage_path
```

Comment:
If you use a custom pyarrow filesystem, then the user needs to pass in an already stripped path, rather than a URI.

```python
(storage_filesystem=CustomS3FileSystem(), storage_path="bucket/a/b/c")
# Cannot use "s3://bucket/a/b/c", since pyarrow.fs.copy_files will complain about a URI being present
```

The default case uses `pyarrow.fs.from_uri` to strip the prefix from the path:

```python
(storage_filesystem=None, storage_path="s3://bucket/a/b/c")
# from_uri(storage_path) -> (DefaultS3FileSystem(), "bucket/a/b/c")
```

Reply:
Question: Is this the desired way to specify a storage path, in the case of a custom filesystem?

Reply:
Does normalize_path() fix this? https://arrow.apache.org/docs/python/generated/pyarrow.fs.FileSystem.html#pyarrow.fs.FileSystem.normalize_path

Not sure I see any other APIs in arrow that would resolve the URI relative to a filesystem.

Reply:
Another option is we can parse URIs and strip the prefix in the case of a custom storage filesystem being passed in.

Reply:
Yeah I tried normalize and it doesn't strip the prefix.

I think stripping the prefix for the user with some urllib parse probably works in most cases, but there are some edge cases. For example, for HDFS: https://github.com/ray-project/ray/blob/6ac4521f2a666b1dff8912f455ac928cbe91c5f4/python/ray/air/_internal/remote_storage.py#L367-L378

Reply:
I see, probably safest to not doing any special parsing then. We can document that users should not specify the URI prefix when using a custom storage filesystem.